[docs] Add a guide on controlled components (#47618)
expo/expo: docs/pages/guides/controlled-components.mdx

---
title: Controlled components
description: Learn how to use controlled components in React Native to control a TextInput, format text as app users type, restrict input, and keep the cursor stable.
---

import { Collapsible } from '~/ui/components/Collapsible';
import { UpdateLoopDiagram } from '~/scenes/guides/controlled-components/UpdateLoopDiagram';
import { WorkletPathDiagram } from '~/scenes/guides/controlled-components/WorkletPathDiagram';

> Every input has a current value, such as the text in an input field, the position of a switch, or the selection in a picker. A **controlled component** keeps the value in your component's state. The input displays whatever the state holds and every change the user makes updates that state. That state is the single source of truth.

Reviewer: I read this sentence as the state has to live in *that* component.

```suggestion
Every input has a current value, such as the text in an input field, the position of a switch, or the selection in a picker. A **controlled component** keeps the value in state you manage. The input displays whatever the state holds and every change the user makes updates that state. That state is the single source of truth.
```

The following example keeps a name in state. `value` displays the state, and [`onChangeText`](https://reactnative.dev/docs/textinput#onchangetext) updates it on every keystroke:

```tsx
import { useState } from 'react';
import { TextInput } from 'react-native';

export default function NameInput() {
  const [name, setName] = useState('');

  return <TextInput value={name} onChangeText={setName} placeholder="Name" />;
}
```

In React Native, this update loop is asynchronous: each update round-trips through the JavaScript thread. A keystroke updates the native field first and then fires `onChangeText`. Your state update triggers a re-render, and the new value travels back to the native field. On the web, React reconciles inputs synchronously against the DOM. Most controlled input problems trace back to that round-trip.

The following diagram shows how a keystroke travels through the update loop:

<UpdateLoopDiagram />

<Collapsible summary="Differences from controlled inputs on web">

If you have written [controlled inputs for the web](https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable), three things change in React Native:

> - The change handler is `onChangeText` and receives the new text directly instead of an event object.

Reviewer: `<TextInput />` has onChange too

```
<TextInput
  onChange={(ev) => setFoo(ev.nativeEvent.text)}
  style={{ borderColor: "red", borderWidth: 1, width: 100, height: 32 }}
  placeholder="Foo"
  value={foo}
/>
```

```suggestion
- The usual change handler is `onChangeText` and receives the new text directly instead of an event object.
```

```

Setting [`keyboardType`](https://reactnative.dev/docs/textinput#keyboardtype) (for example, `number-pad`, `decimal-pad`, or `phone-pad`) shows a matching keyboard, but it does not enforce anything: hardware keyboards and pasted text can still insert other characters, so keep the `onChangeText` filter.

### Force uppercase input

Forcing uppercase is the same pattern as formatting. Pair the transform with [`autoCapitalize`](https://reactnative.dev/docs/textinput#autocapitalize) so the on-screen keyboard produces capital letters to begin with.

The following example stores a coupon code in uppercase:

```tsx
<TextInput
  value={code}
  onChangeText={text => setCode(text.toUpperCase())}
  autoCapitalize="characters"
  placeholder="COUPON"
/>
```

### Keep the cursor in place while formatting

Before React Native's [New Architecture](/guides/new-architecture/), reformatting the value inside `onChangeText` snapped the cursor to the end of the field, which made edits in the middle of the input difficult.

In React Native and Expo apps that use the New Architecture, inserting text in the middle of the field keeps the cursor right after the inserted text.

> If the cursor still ends up in the wrong place (on an older version of React Native, or with a transform that inserts or removes characters before the cursor):

Reviewer: This sounds like a guarantee
> In React Native and Expo apps that use the New Architecture, inserting text in the middle of the field keeps the cursor right after the inserted text.

And the next sentence begins like a contradiction:
> If the cursor still ends up in the wrong place

Author: Good catch! The cursor can still move on the New Architecture when `onChangeText` returns different text than what the user typed. @entiendoNull I've updated this to not self-contradict each other here: https://github.com/expo/expo/pull/47618/changes#diff-ee5a5e92b8c521c5b77904c04a54242a5f4d637043dadc0d9c593ed180b3ffddR166

Reviewer: Looks great!
